### DISCUSSION TITLE
OpenWebUI system to my own CRM system database

### DISCUSSION THREAD

**Discussion by peuportier** [open-webui/open-webui#3898] [Q&A]
Hello fellow developers,

I am exploring the possibility of connecting the OpenWebUI system to my own CRM system database. Specifically, I would like to know if it is feasible to:

Enable the OpenWebUI to access and interact with all tables and data within my CRM database.
Allow real-time analytics through an LLM (Large Language Model) to analyze and provide insights based on the data.
Facilitate the ability to track production line information and other relevant data.
Support the creation and updating of information within the CRM system directly through the OpenWebUI interface.
This integration would significantly enhance our data management and operational efficiency. Any guidance or suggestions on how to achieve this would be greatly appreciated. Pipeline or else ? 

Thank you!

**Comment by andreybond** [comment]
I think you should look into the pipelines.

You basically can implement a sort of a chatbot with some commands supported and allow users to chat with your CRM, of course this implies development of a corresponding logic / workflows within your pipeline(s). So all logic related to the access to your data will be implemented in your custom pipeline.

**Reply by peuportier** [reply]
First, thank you for your answer. My initial idea is to embed OpenWebUI into an iframe that will appear on our CRM interface.

Next, I plan to create a pipeline that connects to our database, allowing the LLM models from Ollama to read the database data. Alternatively, it might be a good approach to first connect the database and see if I can make OpenWebUI understand and interact with my database directly.

What do you think about this approach? Do you have any specific recommendations or best practices for setting up the database connection and integrating it with OpenWebUI? Also, any advice on embedding OpenWebUI effectively would be appreciated. Thanks you and good day

**Reply by peuportier** [reply]
I try first to see if it can connect my DB . make a function as filter. But I try it. Said doesn't work: Maybe you have a clue ? Using the openwebui template. 
```

"""
title: CRM Database Query Pipeline
author: Peuportier Aurelien
author_url: 
funding_url: 
version: 0.1
"""

from pydantic import BaseModel, Field
from typing import Optional
import mysql.connector
from utils.misc import get_last_user_message


class Filter:
    class Valves(BaseModel):
        priority: int = Field(
            default=0, description="Priority level for the filter operations."
        )
        max_turns: int = Field(
            default=8, description="Maximum allowable conversation turns for a user."
        )

    class UserValves(BaseModel):
        max_turns: int = Field(
            default=4, description="Maximum allowable conversation turns for a user."
        )

    def __init__(self):
        self.valves = self.Valves()
        self.connection = self.connect_to_db()

    def connect_to_db(self):
        try:
            connection = mysql.connector.connect(
                host="your_host",
                user="your_user",
                password="your_password",
                database="your_database",
            )
            return connection
        except mysql.connector.Error as err:
            print(f"Error: {err}")
            return None

    def get_all_tables(self):
        if self.connection:
            try:
                cursor = self.connection.cursor()
                cursor.execute("SHOW TABLES")
                tables = cursor.fetchall()
                cursor.close()
                return [table[0] for table in tables]
            except mysql.connector.Error as err:
                print(f"Error: {err}")
                return []
        else:
            return []

    def inlet(self, body: dict, __user__: Optional[dict] = None) -> dict:
        print(f"inlet:{__name__}")
        print(f"inlet:body:{body}")
        print(f"inlet:user:{__user__}")

        user_message = get_last_user_message(body).strip().lower()
        if user_message == "show all tables":
            tables = self.get_all_tables()
            if tables:
                response = f"Tables in the database: {', '.join(tables)}"
            else:
                response = "No tables found or error retrieving tables."
            body["messages"].append({"role": "system", "content": response})

        return body

    def outlet(self, body: dict, __user__: Optional[dict] = None) -> dict:
        print(f"outlet:{__name__}")
        print(f"outlet:body:{body}")
        print(f"outlet:user:{__user__}")

        return body
```
Perhaps it would be a good step to first connect my own database and display all the tables in the chat when I ask specific queries. Using specific keywords to trigger this functionality would be a great starting point.

**Reply by OM-EL** [reply]
i have a question pelease , these two variables will be initialized at the begining when creating the tool.
```
    class Valves(BaseModel):
        priority: int = Field(
            default=0, description="Priority level for the filter operations."
        )
        max_turns: int = Field(
            default=8, description="Maximum allowable conversation turns for a user."
        )
```

        
When will these variable initialized ? 

```

    class UserValves(BaseModel):
        max_turns: int = Field(
            default=4, description="Maximum allowable conversation turns for a user."
        )
```

**Reply by peuportier** [reply]
Hi,

I need to delve deeper into this. For now, I'm trying to get OpenWebUI to connect to my database and simply respond with "yes" or "no" when it is connected. I attempted to achieve this using Tools, but I haven't succeeded yet. Essentially, I want to be able to ask, "Can you tell me if the database is connected?" and have the chat respond with either "Yes, I can connect" or "No, I can't."

This is my first step. Do you have any suggestions? Also, I forgot to delete these variables as I was using a template to write the script.

**Reply by scalenow** [reply]
@peuportier You need to define your business logic first to interact with the db,None of the LLM can create business logic(sql to interact with the Db).I have built using RASA community version bu finalising SSO with keycloak ,once finalised wilk share the architecture

**Reply by peuportier** [reply]
@scalenow 

Thanks you , Would really appreciate the thoughts.  I will have a look to RASA and starts from here. Wish you a good day

**Reply by scalenow** [reply]
@peuportier download the community version on github ,paid versions are very expensive

**Reply by peuportier** [reply]
Got it. Thank you. We need to conduct research to make OWUI more flexible. What do you think about OWUI? A couple of functions or tools would be nice. I'll work on one to connect all Mistral models and use the Vision model through a stream. I'm also trying to figure out how to generate an image from the stream. If you need this function, I'll share it with you :-)

**Reply by JatenJ** [reply]
Please can you share how you solved it? I have the same issue and I am struggling to connect to my database. Can you please share your solution?

**Comment by bmc0923** [comment]
I think there are some misunderstandings here. You should create a mini python MCP server which connects to your CRM API, and then connect the MCP server to OpenWebUI. You could also create a postgres db, and then pull in all data from you CRM on a schedule which might make more sense if you're doing a lot of queries, and data isn't time sensitive. Either way, a pipeline is only needed if you're trying to connect to a different model, or you want to set this up using a custom langgraph backend - which could be good option, but that is not what is needed based on the above. 

I have connected OpenWebUI to ERPs, CRMs, Cloud Storage amongst other services.
